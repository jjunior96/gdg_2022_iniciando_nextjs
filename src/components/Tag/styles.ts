import { lighten } from 'polished';
import styled, { css, DefaultTheme } from 'styled-components';

import { TagProps } from '.';

type ContainerProps = Pick<TagProps, 'color'>;

const containerModifiers = {
  blue: (theme: DefaultTheme) => css`
    background-color: ${lighten(0.34, theme.colors.info)};
    border-color: ${theme.colors.info};

    color: ${theme.colors.info};
  `,

  red: (theme: DefaultTheme) => css`
    background-color: ${lighten(0.34, theme.colors.red)};
    border-color: ${theme.colors.red};

    color: ${theme.colors.red};
  `,

  green: (theme: DefaultTheme) => css`
    background-color: ${lighten(0.34, theme.colors.green)};
    border-color: ${theme.colors.green};

    color: ${theme.colors.green};
  `
};

export const Container = styled.div<ContainerProps>`
  ${({ theme, color }) => css`
    width: fit-content;
    max-width: 20rem;

    padding: 0.5rem ${theme.spacings.small};
    color: ${theme.colors.white};

    font-size: calc(${theme.font.sizes.xxxsmall} - 0.1rem);
    font-weight: ${theme.font.bold};

    border-style: solid;
    border-width: 1px;

    border-radius: 20px;

    overflow: hidden;

    ${!!color && containerModifiers[color!](theme)};
  `}
`;
