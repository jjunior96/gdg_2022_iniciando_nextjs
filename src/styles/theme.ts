export default {
  grid: {
    container: '96rem',
    gutter: '3.2rem'
  },
  components: {
    sidebar: '20rem',
    header: '8rem'
  },
  border: {
    radius: '0.4rem'
  },
  box: {
    shadow: '0 8px 8px rgba(0, 0, 0, 0.2)'
  },
  font: {
    family:
      "Montserrat, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xxxxsmall: '0.8rem',
      xxxsmall: '1.2rem',
      xxsmall: '1.4rem',
      xsmall: '1.6rem',
      small: '2rem',
      medium: '2.4rem',
      large: '3.2rem',
      xlarge: '4rem',
      xxlarge: '4.8rem',
      xxxlarge: '5.6rem',
      huge: '6.4rem'
    }
  },
  colors: {
    primary: '#24A66B',
    secundary: '#5855E9',
    white: '#FAFAFA',
    gray_50: '#f0efeb',
    gray_100: '#D9D9D9',
    gray_200: '#91AAB4',
    gray_300: '#3E606F',
    gray_450: '#193441',
    gray_500: '#343a40',
    yellow_300: '#F1C40F',
    purple_100: '#7F57DF',
    purple_300: '#5855E9',
    green: '#24A66B',
    red: '#E63737',
    success: '#24A66B',
    error: '#E63737',
    info: '#189CEF',
    black: '#0D0D0D'
  },
  spacings: {
    xxxsmall: '0.8rem',
    xxsmall: '1rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  modal: {
    sizes: {
      small: '40rem',
      medium: '48rem',
      large: '54rem'
    }
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const;
