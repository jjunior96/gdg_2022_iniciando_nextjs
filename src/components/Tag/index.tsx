import * as S from './styles';

export type TagColorsProps = 'red' | 'green' | 'blue';

export type TagProps = {
  children: React.ReactNode;
  color?: TagColorsProps;
};

const Tag = ({ children, color = 'blue' }: TagProps) => {
  return (
    <S.Container role="textbox" color={color}>
      {children}
    </S.Container>
  );
};

export default Tag;
