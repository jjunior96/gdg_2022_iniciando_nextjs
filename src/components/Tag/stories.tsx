import { Story, Meta } from '@storybook/react/types-6-0';

import Tag, { TagProps } from '.';

export default {
  title: 'Tag',
  component: Tag
} as Meta;

export const Default: Story<TagProps> = (args) => <Tag {...args} />;

Default.args = {
  children: 'Tag example'
};
