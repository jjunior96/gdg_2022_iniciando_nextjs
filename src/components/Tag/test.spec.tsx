import { screen } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import theme from 'styles/theme';

import Tag from '.';

describe('<Tag />', () => {
  it('should render component by default', () => {
    const { container } = renderWithTheme(<Tag>Tag example</Tag>);

    const tagElement = screen.getByRole('textbox');
    const labelText = screen.getByText(/tag example/i);

    expect(tagElement).toBeInTheDocument();
    expect(tagElement).toHaveStyle({
      'border-color': theme.colors.info,
      color: theme.colors.info
    });

    expect(labelText).toBeInTheDocument();
    expect(labelText).toHaveStyle({
      color: theme.colors.info
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render component with color red', () => {
    renderWithTheme(<Tag color="red">Tag example</Tag>);

    const tagElement = screen.getByRole('textbox');
    const labelText = screen.getByText(/tag example/i);

    expect(tagElement).toHaveStyle({
      'border-color': theme.colors.red,
      color: theme.colors.red
    });

    expect(labelText).toHaveStyle({
      color: theme.colors.red
    });
  });

  it('should render component with color green', () => {
    renderWithTheme(<Tag color="green">Tag example</Tag>);

    const tagElement = screen.getByRole('textbox');
    const labelText = screen.getByText(/tag example/i);

    expect(tagElement).toHaveStyle({
      'border-color': theme.colors.green,
      color: theme.colors.green
    });

    expect(labelText).toHaveStyle({
      color: theme.colors.green
    });
  });
});
