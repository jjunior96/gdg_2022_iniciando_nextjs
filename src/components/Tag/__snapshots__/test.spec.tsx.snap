// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> should render component by default 1`] = `
.c0 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  max-width: 20rem;
  padding: 0.5rem 2.4rem;
  color: #FAFAFA;
  font-size: calc(1.2rem - 0.1rem);
  font-weight: 600;
  border-style: solid;
  border-width: 1px;
  border-radius: 20px;
  overflow: hidden;
  background-color: #bae2fa;
  border-color: #189CEF;
  color: #189CEF;
}

<div
  class="c0"
  color="blue"
  role="textbox"
>
  Tag example
</div>
`;
